pyramid/scripts/pserve.py

```diff
@@ -887,11 +889,28 @@ def check_reload(self):
                 continue
             if filename.endswith('.pyc') and os.path.exists(filename[:-1]):
                 mtime = max(os.stat(filename[:-1]).st_mtime, mtime)
+                pyc = True
+            else:
+                pyc = False
             if filename not in self.module_mtimes:
                 self.module_mtimes[filename] = mtime
             elif self.module_mtimes[filename] < mtime:
-                print("%s changed; reloading..." % filename)
-                return False
+                if pyc:
+                    filename = filename[:-1]
+                # check if a file has syntax errors. If so, track it until it's fixed
+                for filename_promise in self.syntax_error_files + [filename]:
+                    try:
+                        py_compile.compile(filename_promise, doraise=True)
+                    except py_compile.PyCompileError:
+                        print("%s has a SyntaxError; NOT reloading." % filename_promise)
```

Comment:
> `print` is wrong but so is logging,

Why is logging wrong? Logging is user-configurable and allows you to agnostically write to the user's preferred output. As far as informational versus erroneous conditions, that why log levels exist -- you can use `info()` and `error()`.

> For this reason I'm leaning toward showing all the errors all the time.

I think if you do it this way, you're likely to inure the user to important messages. Additionally, what other output would the server generate if it's stopped due to a `SyntaxError`? More messages from other errors.

The key point is that the software should only be generating output in response to user-concerning events. The fact that a loop completed another cycle is not of concern to the user. If it prints out a list of syntactically-incorrect files every N seconds, that's leaking a fact about the server which I, as a user, don't care about.

Additionally, there may be error messages on screen which I *do* care about and need to track *e.g.* stack traces from previous unhandled exceptions. Why should those get bumped out of history by messages being generated on an interval which I don't control?

Suppose a project has `x.py` and `y.py` in it. As a dev,

- If I introduce an error into `x.py`, I would like to see:
  ```
  x.py has an error
  ```
  That error should be printed out once.

- If I then introduce an error into `y.py`, then I would like to see:
  ```
  x.py has an error
  y.py has an error
  ```
  In this case I am learning that my modification to `y.py` triggered a reload cycle, **which I expected**, and additionally `x.py` remains unfixed. These are both germane to my user-initiated action, and therefore I care about seeing the first message re-printed.

- If I then fix `x.py` and save, this triggers another reload, which again is expected. I therefore would expect to see the current list of erroneous files (only `y.py` now)

In short, bothering the user by default feels like a poorly thought out process. The user is liable to feel that they are bearing the cost of the programmer's failure to keep count. I can't think of single situation where someone would see a wall of 
```
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
x.py has a syntax error
...
```
and not find that to be annoying.